### PR TITLE
ref(rules): Fix typing for rule_snooze & rule

### PR DESCRIPTION
**Pull request by Christinarlong**
Trying a new thing where we try and do all (read: most) of the typing upfront and then move the modules to their new home.

issue ref #73858

### CODE REVIEW

File: src/sentry/api/serializers/rest_framework/rule.py [hunk 1]
```diff
@@ -168,31 +177,6 @@ def validate_conditions(self, conditions):
     def validate(self, attrs):
         return super().validate(validate_actions(attrs))
 
-    def save(self, rule):
```
Comment (Christinarlong) [inline src/sentry/api/serializers/rest_framework/rule.py:171]: No code ran/referenced this ._.

File: src/sentry/api/endpoints/rule_snooze.py [hunk 2]
```diff
@@ -104,20 +106,25 @@ def post(self, request: Request, project: Project, rule: Rule | AlertRule) -> Re
 
         if not can_edit_alert_rule(project.organization, request):
             raise PermissionDenied(
-                detail="Requesting user cannot mute this rule.", code=status.HTTP_403_FORBIDDEN
+                detail="Requesting user cannot mute this rule.", code=str(status.HTTP_403_FORBIDDEN)
```
Comment (asottile-sentry) [inline src/sentry/api/endpoints/rule_snooze.py:109]: the `str(403)` seems suspicious -- what was this fixing? do we even need `code=` (surely a PermissionDenied is always going to be a 403 right?)

File: src/sentry/api/endpoints/rule_snooze.py [hunk 3]
```diff
@@ -76,14 +77,15 @@ def serialize(self, obj, attrs, user, **kwargs):
 @region_silo_endpoint
 class BaseRuleSnoozeEndpoint(ProjectEndpoint):
     permission_classes = (ProjectAlertRulePermission,)
-    rule_model = type[Rule] | type[AlertRule]
+    rule_model: type[AlertRule | Rule]
```
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/api/endpoints/rule_snooze.py:80]: might be useful to have an `# abstract` comment on this just so it's clear

btw the original code was a type alias so I'm surprised this worked at all!

File: src/sentry/api/serializers/rest_framework/rule.py (hunk 1 above)
Comment (Christinarlong) [reply src/sentry/api/serializers/rest_framework/rule.py:171]: 'o' is `.save()` run implicitly when people instantiate the serializer ? I couldn't find any explicit calls for `.save()`, and this method doesn't implement any of the `create()` or `update()`. So, we should be safe ?

File: src/sentry/api/serializers/rest_framework/rule.py [hunk 4]
```diff
@@ -64,10 +69,15 @@ def to_internal_value(self, data):
             # XXX(epurkhiser): Very hacky, but we really just want validation
             # errors that are more specific, not just 'this wasn't filled out',
             # give a more generic error for those.
-            first_error = next(iter(form.errors.values()))[0]
```
[the remark this answers is not in the document]
Comment (Christinarlong) [reply src/sentry/api/serializers/rest_framework/rule.py:67]: yesssss, though the `.values()` transformation makes the result a `UserList[_StrOrPromise]` in reality, so I guess we could just type hint the variable to that.

File: src/sentry/api/endpoints/rule_snooze.py (hunk 2 above)
Comment (Christinarlong) [reply src/sentry/api/endpoints/rule_snooze.py:109]: yeah, good pt. the 403 comes included in the `PermissionDenied` exception class already

File: src/sentry/api/serializers/rest_framework/rule.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/api/serializers/rest_framework/rule.py:67]: I would just remove the annotation personally

File: src/sentry/api/serializers/rest_framework/rule.py (hunk 1 above)
Comment (asottile-sentry) [reply src/sentry/api/serializers/rest_framework/rule.py:171]: I am not sure -- I thought it was magically invoked by forms somehow -- could probably trace when this function was introduced and see how it was used in that PR and then from there figure out if that code path is still used

Comment (Christinarlong) [reply src/sentry/api/serializers/rest_framework/rule.py:171]: :0) good thinking, looks like this (vintage) function used to be used in the rules endpoints, but now we use dataclasses (see: [ProjectRuleCreator](https://github.com/getsentry/sentry/blob/master/src/sentry/projects/project_rules/creator.py) & [ProjectRuleUpdater](https://github.com/getsentry/sentry/blob/master/src/sentry/projects/project_rules/updater.py#L13)) that do the same actions instead

File: src/sentry/api/endpoints/rule_snooze.py [hunk 5]
```diff
@@ -158,9 +165,14 @@ def delete(self, request: Request, project: Project, rule: Rule | AlertRule) ->
         # find if there is a mute for all that I can remove
         shared_snooze = None
         deletion_type = None
-        kwargs = {self.rule_field: rule, "user_id": None}
         try:
-            shared_snooze = RuleSnooze.objects.get(**kwargs)
+            # Due to mypy requiring keyword arguments to not be kwargs, we need to check the type of rule and query accordingly
+            # In RuleSnooze, rule and alert_rule are mutually exclusive (cannot both be not null)
```
[the remark this answers is not in the document]
Comment (Christinarlong) [reply src/sentry/api/endpoints/rule_snooze.py:170]: nvm 🥲 it turns out a union of Literals in a variable key for a typed dict isn't supported by mypy currently (https://github.com/python/mypy/issues/7752)

```python
class RuleField(TypedDict, total=False):
     rule: Rule
     alert_rule: AlertRule

kwargs: RuleField = {self.rule_field: rule}
```

[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/api/endpoints/rule_snooze.py:170]: really starting to think this base class is a maintenance problem and trying to be too cute -- and that these are actually two disparate things trying to be shoved into a BaseRuleSnoozeEndpoint shaped box

[the remark this answers is not in the document]
Comment (Christinarlong) [reply src/sentry/api/endpoints/rule_snooze.py:170]: yeahhhhh like we could delegate the RuleSnooze call to the child classes, and keep the rest of the logic in a helper in the base case.

[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/api/endpoints/rule_snooze.py:170]: yeah or a few composable functions would probably work too!

[the remark this answers is not in the document]
Comment (Christinarlong) [reply src/sentry/api/endpoints/rule_snooze.py:170]: Wait, can you elaborate more on what composable functions means in this context ? The google tells me composable functions when we chain functions together/when functions take in other functions but I'm not sure how that would help in this case :thonk

[the remark this answers is not in the document]
Comment (asottile-sentry) [reply src/sentry/api/endpoints/rule_snooze.py:170]: maybe won't need to compose -- but can probably make two separate functions for `create_rule_snooze` and `delete_rule_snooze` and then call them with the appropriate arguments in the two "subclasses" here -- basically undo the inheritance and instead have free functions which accomplish the same thing (but are less reliant on weird conventions of `self` to shoehorn in the right arguments).

[the remark this answers is not in the document]
Comment (markstory) [reply src/sentry/api/endpoints/rule_snooze.py:170]: Pushing the differences in the subclasses into methods/functions could work. We have similar patterns for the `OrganizationUnsubscribe` endpoints and `AvatarMixin`.

**Review by asottile-sentry — APPROVED** [review]

**Pull request merged**